File: src-tauri/src/infra/path_utils.rs
use std::path::Path;

pub fn sanitize_project_directory_name(name: &str) -> String {
    let trimmed = name.trim();
    let mut no_illegal_chars = String::with_capacity(trimmed.len());
    for ch in trimmed.chars() {
        if ch.is_control() || matches!(ch, '<' | '>' | ':' | '"' | '/' | '\\' | '|' | '?' | '*') {
            no_illegal_chars.push('_');
        } else {
            no_illegal_chars.push(ch);
        }
    }

    let collapsed = no_illegal_chars
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    if collapsed.is_empty() {
        "novelforge-project".to_string()
    } else {
        collapsed
    }
}

pub fn chapter_file_name(index: i64) -> String {
    format!("ch-{index:04}.md")
}

pub fn to_posix_relative(from_dir: &Path, target: &Path) -> String {
    match target.strip_prefix(from_dir) {
        Ok(relative) => {
            let mut parts = Vec::new();
            for part in relative.components() {
                parts.push(part.as_os_str().to_string_lossy().into_owned());
            }
            parts.join("/")
        }
        Err(_) => target.to_string_lossy().replace('\\', "/"),
    }
}

File: tests/dev-engine-fixtures/infra/path-utils.ts
import path from "node:path";

export function toPosixRelative(fromDir: string, target: string): string {
  return path.relative(fromDir, target).split(path.sep).join("/");
}

export function sanitizeProjectDirectoryName(name: string): string {
  const trimmed = name.trim();
  const noIllegalChars = trimmed.replace(/[<>:"/\\|?*\x00-\x1F]/g, "_");
  const collapsed = noIllegalChars.replace(/\s+/g, " ").trim();
  return collapsed.length > 0 ? collapsed : "novelforge-project";
}

export function chapterFileName(index: number): string {
  return `ch-${index.toString().padStart(4, "0")}.md`;
}
